HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift

```diff
@@ -15,45 +15,248 @@
 import XCTest
 @testable import HeartbeatLogging
 
-// TODO: Add additional validation (#8896 comments).
+// TODO: Unit test recording across time zones
 
 class HeartbeatControllerTests: XCTestCase {
+  // 2021-11-01 @ 00:00:00 (EST)
+  let date = Date(timeIntervalSince1970: 1_635_739_200)
+
   func testFlushWhenEmpty() throws {
     // Given
     let controller = HeartbeatController(storage: HeartbeatStorageFake())
+    // Then
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+  }
+
+  func testLogAndFlush() throws {
+    // Given
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { self.date }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
     // When
-    let flushed = controller.flush()
+    controller.log("dummy_agent")
+    let heartbeatPayload = controller.flush()
+
     // Then
-    XCTAssertNil(flushed)
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
   }
 
-  func testLogThenFlush() throws {
+  func testLoggingDifferentAgentsInSameTimePeriodOnlyStoresTheFirst() throws {
     // Given
-    let controller = HeartbeatController(storage: HeartbeatStorageFake())
-    XCTAssertNil(controller.flush())
+    let testDate = date
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { testDate }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    controller.log("dummy_agent")
+    controller.log("some_other_dummy_agent")
+    let heartbeatPayload = controller.flush()
+
+    // Then
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+  }
+
+  func testLogAtEndOfTimePeriodAndAcceptAtStartOfNextOne() throws {
+    // Given
+    var testDate = date
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { testDate }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    // - Clock time 2021-11-01 @ 00:00:00 (EST)
+    controller.log("dummy_agent")
+
+    // - Advance to 2021-11-01 @ 23:59:59 (EST)
+    testDate.addTimeInterval(60 * 60 * 24 - 1)
```

Comment:
Thanks to your comments @maksymmalyhin, I took a closer look and realized I was able to remove the `SystemTime` class entirely because the `HeartbeatController` initializer captures the `testDate` in an `@escaping` closure so we can treat the date with reference semantics (`testDate.addTimeInterval(60 * 60 * 24 - 1)`) and the controller's date provider will reflect those changes when we use it in public API implementation.